### DISCUSSION TITLE
CBS pipeline might not result in peak performance

### DISCUSSION THREAD

**Discussion by ravil-mobile** [AdaptiveCpp/AdaptiveCpp#783] [General]
Hi all, 

I am compiling my application with hipSycl. Everything works fine. However, I am constantly getting the following warning:
```
[hipSYCL Warning] CBS pipeline might not result in peak performance with old PM
```

I cannot get what exactly this warning means. What are `CBS` and `PM`? Would you mind explaining these terms?

**Comment by illuhad** [comment]
Hi!

The CBS pipeline is our implementation of compiler acceleration for `nd_range` parallel for kernels on CPU devices. It's what makes `nd_range` kernels fast when running on CPU targets.

The pass manager, `PM`, is part of the the clang infrastructure.

What this warning tells you is that the CPU kernel optimizations might be suboptimal due to it using the pass manager infrastructure of older clang versions. This can be fixed by upgrading your clang. @fodinabor  can give you details.

If you are not running `nd_range` kernels on CPU, you can ignore this warning, as it will exclusively affect kernel performance on the CPU device.

**Reply by ravil-mobile** [reply]
Hi, 

Thanks for a quick reply!

> If you are not running nd_range kernels on CPU, you can ignore this warning

No, I am only using your GPU backends.

>  This can be fixed by upgrading your clang.

I am curious about which version of `clang` compiler you recommend. I've been using clang@11 and clang@12 for `hipsycl`.

I have one more question related to `gpu_selector` i.e., how to pick up the same device as is listed in `nvidia_visible_devices`. It is related to a multiple GPU installations of compute nodes. Shall I open a new discussion?

**Reply by illuhad** [reply]
> I am curious about which version of clang compiler you recommend. I've been using clang@11 and clang@12 for hipsycl.

We currently support clang/LLVM 11-14. Generally, newer LLVM versions perform better due to better optimizations or provide more functionality. I think the new pass manager is a clang 13 thing. Will you see a difference for your use case? I don't know.
If you just want to get rid of the warning and don't care about performance on CPU, you can either disable CBS support by building hipSYCL with cmake option `-DWITH_ACCELERATED_CPU=OFF`, or just using `--hipsycl-targets="omp.library-only"` instead of just targeting `omp` or `omp.accelerated`.

> I have one more question related to gpu_selector i.e., how to pick up the same device as is listed in nvidia_visible_devices. It is related to a multiple GPU installations of compute nodes. Shall I open a new discussion?

I'm not sure what you mean by `nvidia_visible_devices`. Do you mean the `NVIDIA_VISIBLE_DEVICES/CUDA_VISIBLE_DEVICES` environent variables? hipSYCL just uses the regular CUDA API, so anything that changes the visiblity of CUDA devices will also change what hipSYCL sees, and therefore, what `gpu_selector` can return.

**Reply by ravil-mobile** [reply]
@illuhad, Thanks. `-DWITH_ACCELERATED_CPU=OFF` help.

> I'm not sure what you mean by nvidia_visible_devices. Do you mean the NVIDIA_VISIBLE_DEVICES/CUDA_VISIBLE_DEVICES environent variables? hipSYCL just uses the regular CUDA API, so anything that changes the visiblity of CUDA devices will also change what hipSYCL sees, and therefore, what gpu_selector can return.

We have a mixed CUDA/HIP + OpenMP application. It happened because of some historical reasons. In short, we use `cuda/rocm` to get the best possible performance for the generated code (we're using code generation in our project) and OpenMP offloading for the rest of the code in order to achieve portability. However, I am not happy with OpenMP because it results in low GPU performance when I am using `clang/gcc` compilers. The current idea is to substitute OpenMP with SYCL 2020. 

I checked CUDA+SYCL (with `hipsycl`) and it worked with a single GPU setup. However, I am mostly targeting multi-GPU per node setups. To make it work, I need to be sure that each each MPI process picks the same GPU instance in both CUDA and hipSYCL.

My CUDA part is using `cudaSetDevice(localMPIRank)`  where `localMPIRank` is an integer number. In SYCL, I can do the following:
```
    auto platform = sycl::platform(sycl::gpu_selector());
    auto device = platform.get_devices()[localMPIRank] ;
```
If I configure my project  in CMake as follows:
```
set(HIPSYCL_TARGETS "cuda:sm_75")
find_package(hipSYCL CONFIG REQUIRED)
```

`hipsycl` is going to return me only those devices which are capable of running on `sm_75` architecture.

The thing is that I am not sure that `cudaSetDevice(localMPIRank)` and `auto device = platform.get_devices()[localMPIRank]` is going to refer to the same GPU.

What do you think?

**Reply by illuhad** [reply]
> If I configure my project in CMake as follows: [...] hipsycl is going to return me only those devices which are capable of running on sm_75 architecture.

No, specifying hipSYCL targets is a compiler thing. It only impacts for which devices the compiler generates code. It does not impact which devices are visible to the hipSYCL runtime. These are two very different things, e.g. you might want to be able to interact with devices for which you do not have compiled SYCL kernels in case you have external libraries that do. E.g., if you just want to do backend interoperability to invoke cuBLAS on these devices.

> The thing is that I am not sure that cudaSetDevice(localMPIRank) and auto device = platform.get_devices()[localMPIRank] is going to refer to the same GPU.

It should be the exact same thing. hipSYCL just uses the same CUDA API under the hood, and it relies on the CUDA device enumeration to identify all CUDA devices. So, it will return them in the same order as CUDA.

**Reply by illuhad** [reply]
Addendum: A more idiomatic approach for this use case might be to use the SYCL 2020 backend interoperability API. With this, you can construct a SYCL device from the native device of the backend - in the case of hipSYCL's CUDA backend, the CUDA device index:

```
int cuda_device_index = localMPIRank;
#include <sycl/sycl.hpp>
// Construct a device from the CUDA backend with native CUDA device id == localMPIRank
sycl::device d = sycl::make_device<sycl::backend::cuda>(cuda_device_index);
```
This will be the same as `cudaSetDevice(localMPIRank)` whenever `d` is used.

If you use hipSYCL and the CUDA runtime API in the same application with multiple GPUs, keep in mind that hipSYCL may have to call `cudaSetDevice()` to operate. This usually has limited impact for the user application, because hipSYCL does this mostly in a background worker thread, so the change to the active device will be limited to this thread.
However, for some cases it might matter:
* The USM memory management functions are synchronous and might invoke `cudaSetDevice()` in the calling thread
* When the hipSYCL runtime starts (typically when the first SYCL thing is done in the application) it may iterate through all devices to construct streams.

**Reply by ravil-mobile** [reply]
Hi,

I tried `sycl::device d = sycl::make_device<sycl::backend::cuda>(cuda_device_index);` and it didn't quite work for me. My application is crashing on a multi-gpu setup saying `invalid resource handle (400)`.

I manage to circumvent it with the following launching script
```
$ cat ./launch
#!/bin/bash

if [[ ! -z "${OMPI_COMM_WORLD_LOCAL_RANK}" ]]; then
  LOCAL="${OMPI_COMM_WORLD_LOCAL_RANK}"

elif [[ ! -z "${MV2_COMM_WORLD_LOCAL_RANK}" ]]; then
  LOCAL="${MV2_COMM_WORLD_LOCAL_RANK}"

elif [[ ! -z "${PMI_RANK}" ]]; then
  LOCAL="${PMI_RANK}"

elif [[ ! -z "${SLURM_LOCALID}" ]]; then
  LOCAL="${SLURM_LOCALID}"

else
  LOCAL=0
fi

export CUDA_VISIBLE_DEVICES="${LOCAL}"
"$@"

$ ./launch myApp ./parameters.par
```

But there is yet another problem i.e., no hipsycl kernels were executed. Here is a piece of my profiling:
![image](https://user-images.githubusercontent.com/19637079/181225653-b252da43-23c0-448a-bf28-d25c89f5c678.png)
 
Interestingly, everything works on my workstation which has a single GPU.

Do you have any idea what can be wrong?

**Reply by illuhad** [reply]
If you set the environment variable `HIPSYCL_DEBUG_LEVEL=3` before starting your application, hipSYCL will print the detected devices at the very start of the program (after that, there will be a lot of other debug output). Can you please verify that hipSYCL sees the devices that you think should be there?

**Reply by ravil-mobile** [reply]
The problem is that all my GPUs are the same. 
```
$ nvidia-smi 
Wed Jul 27 13:34:50 2022       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.63.01    Driver Version: 470.63.01    CUDA Version: 11.4     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ...  On   | 00000000:01:00.0 Off |                  N/A |
| 30%   24C    P8    32W / 320W |      1MiB / 10018MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  NVIDIA GeForce ...  On   | 00000000:41:00.0 Off |                  N/A |
| 30%   24C    P8    28W / 320W |      1MiB / 10018MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   2  NVIDIA GeForce ...  On   | 00000000:81:00.0 Off |                  N/A |
| 30%   26C    P8    31W / 320W |      1MiB / 10018MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   3  NVIDIA GeForce ...  On   | 00000000:C1:00.0 Off |                  N/A |
| 30%   24C    P8    29W / 320W |      1MiB / 10018MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+


$ CUDA_VISIBLE_DEVICES=2 ./SeisSol_RelWithDebInfo_ssm_86_cuda_6_elastic ./parameters.par
Wed Jul 27 13:32:17, Info:  Initialize Friction Model 
ESC[;32m[hipSYCL Info] ESC[0mbackend_loader: Searching path for backend libs: '"/u/home/ga83dit/usr/hipsycl-0.9.2/lib/hipSYCL"'
ESC[;32m[hipSYCL Info] ESC[0mbackend_loader: Successfully opened plugin: "/u/home/ga83dit/usr/hipsycl-0.9.2/lib/hipSYCL/librt-backend-omp.so" for backend 'omp'
ESC[;32m[hipSYCL Info] ESC[0mbackend_loader: Successfully opened plugin: "/u/home/ga83dit/usr/hipsycl-0.9.2/lib/hipSYCL/librt-backend-cuda.so" for backend 'cuda'
ESC[;32m[hipSYCL Info] ESC[0mRegistering backend: 'omp'...
ESC[;32m[hipSYCL Info] ESC[0mmulti_queue_executor: Spawned for backend OpenMP with configuration: 
ESC[;32m[hipSYCL Info] ESC[0m  device 0: 
ESC[;32m[hipSYCL Info] ESC[0m    memcpy lane: 0
ESC[;32m[hipSYCL Info] ESC[0m    kernel lane: 1
ESC[;32m[hipSYCL Info] ESC[0mRegistering backend: 'cuda'...
ESC[;32m[hipSYCL Info] ESC[0mmulti_queue_executor: Spawned for backend CUDA with configuration: 
ESC[;32m[hipSYCL Info] ESC[0m  device 0: 
ESC[;32m[hipSYCL Info] ESC[0m    memcpy lane: 0
ESC[;32m[hipSYCL Info] ESC[0m    memcpy lane: 1
ESC[;32m[hipSYCL Info] ESC[0m    kernel lane: 2
ESC[;32m[hipSYCL Info] ESC[0m    kernel lane: 3
ESC[;32m[hipSYCL Info] ESC[0mDiscovered devices from backend 'OpenMP': 
ESC[;32m[hipSYCL Info] ESC[0m  device 0: 
ESC[;32m[hipSYCL Info] ESC[0m    vendor: the hipSYCL project
ESC[;32m[hipSYCL Info] ESC[0m    name: hipSYCL OpenMP host device
ESC[;32m[hipSYCL Info] ESC[0mDiscovered devices from backend 'CUDA': 
ESC[;32m[hipSYCL Info] ESC[0m  device 0: 
ESC[;32m[hipSYCL Info] ESC[0m    vendor: NVIDIA
ESC[;32m[hipSYCL Info] ESC[0m    name: NVIDIA GeForce RTX 3080
ESC[;32m[hipSYCL Info] ESC[0mdag_manager: DAG manager is alive!
ESC[;32m[hipSYCL Info] ESC[0mruntime: ******* rt launch initiated ********
ESC[;32m[hipSYCL Info] ESC[0mqueue: Constructed queue with node group id 1
```

I also added `cudaSetDevice(currDeviceId)` right after `sycl::make_device<sycl::backend::cuda>(currDeviceId)`. But it didn't help. I am using v0.9.2 tag hipSycl. Shall I try the master branch?

**Reply by illuhad** [reply]
Okay, I think I see the problem. If you use `CUDA_VISIBLE_DEVICES`, it's the same as removing all devices from the system that are not mentioned in the visibility list from the point of view of the CUDA application. If you only have a single device in `CUDA_VISIBLE_DEVICES`, this device will always have the CUDA device index 0, since it is the only one that is actually visible.
I believe this probably is a general CUDA thing, not necessarily hipSYCL-specific.

**Reply by ravil-mobile** [reply]
Hi, I've just compiled and installed hipSYCL from the `develop` branch (`commit a227f8d9fee65`) and everything worked correctly (i.e., I got correct numerical results and nice performance)! Probably something happen between `v0.9.2` tag and `a227f8d9fee65`

**Comment by fodinabor** [comment]
As @illuhad already mentioned, this only affects you if you are using `parallel_for` kernels with `nd_range` on CPU.

Even though you don't use the CPU backend, I just wanted to leave a reference regarding the details:
Clang < 13 by default uses the old Pass Manager (PM), whereas from Clang >= 13 on, the new PM is the default. From Clang 11+ (or so) you can use `-fexperimental-new-pass-manager` to use the new PM anyways (note: it has reasons why it's called experimental for those versions :)).

The reason why this is matters for hipSYCL is that its Clang plugin does add a number of passes to the optimization pipeline (aka to the PM). The way this works with the old PM, it is not possible to add a certain type of passes (module passes).
During my work on the accelerated CPU backend, I found out that it's quite important that a few passes run before the plugin's main transformations happen, to get good quality code generation. One of those is LLVM's IPSCCP, which is a module pass. Since this is not possible to trigger with the old PM from hipSYCL's plugin, the generated code usually is not quite as good. For example, vectorization fails significantly more often due to the code difference. So that's why there's that warning and that's what it tries to warn about :)

**Reply by ravil-mobile** [reply]
Got you. Thanks for explaining!
